### PR TITLE
Add Samples Directory (and WebCodecs Echo Sample)

### PR DESCRIPTION
**Pull request by aboba**
Fixes https://github.com/w3c/webtransport/issues/416

Live webcodecs echo sample: https://webrtc.internaut.com/wc/wtSender2/

### CODE REVIEW

**Comment by aboba** [comment]
@pthatcherg PTAL

File: samples/webcodecs-echo/js/stream_worker.js [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+               const queue = this.decoder.decodeQueueSize;
+               decqueue_update(queue);
+               const before = performance.now();
+               this.decoder.decode(chunk);
+               const after = performance.now();
+               const duration = after - before;
+               dec_update(duration);
+             } catch (e) {
+               self.postMessage({severity: 'fatal', text: 'Derror size: ' + chunk.byteLength + ' seq: ' + chunk.seqNo + ' kf: ' + chunk.keyframeIndex + ' delta: ' + chunk.deltaframeIndex + ' dur: ' + chunk.duration + ' ts: ' + chunk.timestamp + ' ssrc: ' + chunk.ssrc + ' pt: ' + chunk.pt + ' tid: ' + chunk.temporalLayerId + ' type: ' + chunk.type});
+               self.postMessage({severity: 'fatal', text: `Catch Decode error: ${e.message}`});
```
Comment (jan-ivar) [inline samples/webcodecs-echo/js/stream_worker.js:484]: maybe rethrow to error the stream?

File: samples/webcodecs-echo/js/stream_worker.js [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+     });
+   }
+
+   async start()
+   {
+     if (stopped) return;
+     started = true;
+     self.postMessage({text: 'Start method called.'});
+     try { 
+       this.inputStream
```
Comment (jan-ivar) [inline samples/webcodecs-echo/js/stream_worker.js:669]: missing `await` needed to handle stream errors properly.

Strategies to run both this `this.ReceiveStream` below in parallel include using IFFEs, or assigning promise from each to `const p1` and `const p2` respectively and doing a `Promise.all([p1, p2])` (with or without individual error handling).

File: samples/webcodecs-echo/js/stream_worker.js [hunk 3]
```diff
@@ -0,0 +1,685 @@
...
+       }, 
+       async abort(reason) {
+         // called when ws.abort(reason)
+         // close the transport?
+         await transport.close(); 
+       } 
+     });
+   }
+
+   ReceiveStream(self, transport) {
```
Comment (jan-ivar) [inline samples/webcodecs-echo/js/stream_worker.js:630]: Ditto here
```suggestion
   createReceiveStream(self, transport) {
```
?

File: samples/webcodecs-echo/js/stream_worker.js [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+     self.postMessage({severity: 'fatal', text: 'stop() called'});
+     // TODO: There might be a more elegant way of closing a stream, or other
+     // events to listen for.
+     if (encoder.state != "closed") encoder.close();
+     if (decoder.state != "closed") decoder.close();
+     self.postMessage({severity: 'fatal', text: "stop(): frame, encoder and decoder closed"});
+     return;
+   }
+
+   SendStream(self, transport) {
```
Comment (jan-ivar) [inline samples/webcodecs-echo/js/stream_worker.js:602]: I like this pattern, but it's an odd upper-case named function which call sites aren't using as a constructor (e.g. `new this.SendStream(...)`), so how about:
```suggestion
   createSendStream(self, transport) {
```
?

File: samples/webcodecs-echo/js/main.js [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+      // Create a MediaStreamTrackProcessor, which exposes frames from the track
+      // as a ReadableStream of VideoFrames.
+      let [track] = mediaStream.getVideoTracks();
+      let ts = track.getSettings();
+      const processor = new MediaStreamTrackProcessor(track);
+      inputStream = processor.readable;
+
+      // Create a MediaStreamTrackGenerator, which exposes a track from a
+      // WritableStream of VideoFrames.
+      const generator = new MediaStreamTrackGenerator({kind: 'video'});
```
Comment (jan-ivar) [inline samples/webcodecs-echo/js/main.js:220]: Please add a comment that this is using non-standard Chrome-only versions of *MediaStreamTrackProcessor* and *VideoTrackGenerator* for now, and creating them on main-thread not on the worker, due to lack of transferable MediaStreamTrack in any implementation.

Maybe someday I'll attempt a polyfill in adapter.js...

**Review by jan-ivar — CHANGES_REQUESTED** [review]
This is awesome! I just have some nits to make it a bit less implementation specific.

File: samples/webcodecs-echo/index.html [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+
+    <meta charset="utf-8">
+    <meta name="description" content="WebCodecs code samples">
+    <meta name="viewport" content="width=device-width, user-scalable=yes, initial-scale=1, maximum-scale=1">
+    <meta itemprop="description" content="Client-side WebCodecs code samples">
+    <meta itemprop="name" content="WebCodecs code samples">
+    <meta name="mobile-web-app-capable" content="yes">
+    <meta id="theme-color" name="theme-color" content="#ffffff">
+  <!-- WebTransport origin trial token. See https://developer.chrome.com/origintrials/#/view_trial/793759434324049921 -->
+  <meta http-equiv="origin-trial" content="AvGH7ltY9/n6/2J9PFL9JYg2Xe2pMMfZMEyVNQapgiMSSadPTPTii5NHN4WD9qisPj5jXEoGI4Bp4esA5R2xlgsAAABaeyJvcmlnaW4iOiJodHRwczovL3dlYnJ0Yy5pbnRlcm5hdXQuY29tOjQ0MyIsImZlYXR1cmUiOiJXZWJUcmFuc3BvcnQiLCJleHBpcnkiOjE2MzQwODMxOTl9">
```
[the remark this answers is not in the document]
Comment (aboba) [reply samples/webcodecs-echo/index.html:13]: Fixed.

File: samples/webcodecs-echo/js/main.js [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+      });
+    }
+    gotDevices(await navigator.mediaDevices.enumerateDevices());
+    constraints.deviceId = (videoSource ? {exact: videoSource} : undefined);
+    //addToEventLog('videoSource: ' + JSON.stringify(videoSource) + ' DeviceId: ' + constraints.deviceId);
+  } catch(e) {
+    addToEventLog(`EnumerateDevices error: ${e.message}`);
+  }
+  try {
+    mediaStream = await navigator.mediaDevices.getUserMedia(constraints);
```
[the remark this answers is not in the document]
Comment (aboba) [reply samples/webcodecs-echo/js/main.js:111]: Is there a good sample somewhere that illustrates how to do this in an interoperable way?  I based this on a Chrome sample, which is somewhat flaky.

File: samples/webcodecs-echo/js/stream_worker.js (hunk 2 above)
Comment (aboba) [reply samples/webcodecs-echo/js/stream_worker.js:722]: Thanks. Will investigate.....

File: samples/webcodecs-echo/js/stream_worker.js (hunk 3 above)
Comment (aboba) [reply samples/webcodecs-echo/js/stream_worker.js:680]: Fixed.

File: samples/webcodecs-echo/js/stream_worker.js (hunk 4 above)
Comment (aboba) [reply samples/webcodecs-echo/js/stream_worker.js:612]: Fixed.

File: samples/webcodecs-echo/js/main.js (hunk 5 above)
Comment (aboba) [reply samples/webcodecs-echo/js/main.js:220]: Fixed.

File: samples/webcodecs-echo/js/main.js [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+document.addEventListener('DOMContentLoaded', async function(event) {
+  if (stopped) return;
+  addToEventLog('DOM Content Loaded');
+
+  if (typeof MediaStreamTrackProcessor === 'undefined' ||
+      typeof MediaStreamTrackGenerator === 'undefined') {
+    addToEventLog('Your browser does not support the experimental Mediacapture-transform API.\n' +
+        'Please launch with the --enable-blink-features=WebCodecs,MediaStreamInsertableStreams flag','fatal');
```
[the remark this answers is not in the document]
Comment (aboba) [reply samples/webcodecs-echo/js/main.js:155]: Worth discussing whether mediacapture-transform is being used as intended in this demo, which doesn't implement a jitter buffer upstream of MSTG.  To what extent is MSTG expected to be able to handle jank (which could occur due to a variety of reasons other than network effects)?  In the WebCodecs samples, rendering is done using Canvas, WebGL or WebGPU, not MSTG. I wanted to try out MSTG, but am wondering whether it's responsible for some of the uneven rendering.

**Comment by aboba** [comment]
Going to rebase and resubmit the PR (and other samples).

**Pull request closed without merging**